engine/runner/src/main/java/org/enso/runner/Main.java

```diff
@@ -0,0 +1,1413 @@
+            .build();
+
+    var topScope = context.getTopScope();
+    try {
+      topScope.compile(shouldCompileDependencies);
+      throw exitSuccess();
+    } catch (Throwable t) {
+      logger.error("Unexpected internal error", t);
+      throw exitFail();
+    } finally {
+      context.context().close();
+    }
+  }
+
+  /**
+   * Handles the `--run` CLI option.
+   *
+   * <p>If `path` is a directory, so a project is run, a conflicting (pointing to another project)
+   * `projectPath` should not be provided.
+   *
+   * @param path path of the project or file to execute
+   * @param projectPath if specified, the script is run in context of a project located at that path
+   * @param logLevel log level to set for the engine runtime
+   * @param logMasking is the log masking enabled
+   * @param enableIrCaches are IR caches enabled
+   * @param disablePrivateCheck Is private modules check disabled. If yes, `private` keyword is
+   *     ignored.
+   * @param inspect shall inspect option be enabled
+   * @param dump shall graphs be sent to the IGV
+   * @param executionEnvironment optional name of the execution environment to use during execution
+   */
+  private void run(
+      String path,
+      java.util.List<String> additionalArgs,
+      scala.Option<String> projectPath,
+      Level logLevel,
+      boolean logMasking,
+      boolean enableIrCaches,
+      boolean disablePrivateCheck,
+      boolean enableAutoParallelism,
+      boolean inspect,
+      boolean dump,
+      scala.Option<String> executionEnvironment,
+      int warningsLimit)
+      throws IOException {
+    var fileAndProject = Utils.findFileAndProject(path, projectPath.getOrElse(() -> null));
+    if (fileAndProject == null) {
+      throw exitFail();
+    }
+    var projectMode = fileAndProject._1();
+    var file = fileAndProject._2();
+    var projectRoot = fileAndProject._3();
+    var options = new HashMap<String, String>();
+    if (dump) {
+      options.put("engine.TraceCompilation", "true");
+      options.put("engine.MultiTier", "false");
+    }
+    if (inspect) {
+      options.put("inspect", "");
+    }
+    var context =
+        ContextFactory.create()
+            .projectRoot(projectRoot)
+            .in(System.in)
+            .out(System.out)
+            .repl(new Repl(makeTerminalForRepl()))
+            .logLevel(logLevel)
+            .logMasking(logMasking)
+            .enableIrCaches(enableIrCaches)
+            .disablePrivateCheck(disablePrivateCheck)
+            .strictErrors(true)
+            .enableAutoParallelism(enableAutoParallelism)
+            .executionEnvironment(
+                executionEnvironment.isDefined() ? executionEnvironment.get() : null)
+            .warningsLimit(warningsLimit)
+            .options(options)
+            .build();
+
+    if (projectMode) {
+      var result = PackageManager$.MODULE$.Default().loadPackage(file);
+      if (result.isSuccess()) {
```

Comment:
Feels like just going `instanceof` woule be easier to read

Reply:
It tried that first. It was a mess. However it might be caused by me writing the `switch`/`case` statements incorrectly.